wandb/sdk/internal/job_builder.py

```diff
@@ -0,0 +1,252 @@
+                os.path.basename(sys.executable),
+                program_relpath,
+            ],
+            "args": args,
+            "git": {
+                "remote": remote,
+                "commit": commit,
+            },
+        }
+
+        name = make_artifact_name_safe(f"job-{remote}_{program_relpath}")
+
+        artifact = Artifact(name, JOB_ARTIFACT_TYPE)
+        if os.path.exists(os.path.join(self._settings.files_dir, DIFF_FNAME)):
+            artifact.add_file(
+                os.path.join(self._settings.files_dir, DIFF_FNAME),
+                name="diff.patch",
+            )
+        return artifact, source
+
+    def _build_artifact_job(
+        self, program_relpath: str, args: List[str]
+    ) -> Tuple[Artifact, ArtifactSourceDict]:
+        assert isinstance(self._logged_code_artifact, dict)
+        source: ArtifactSourceDict = {
+            "entrypoint": [
+                os.path.basename(sys.executable),
+                program_relpath,
+            ],
+            "args": args,
+            "artifact": f"wandb-artifact://_id/{self._logged_code_artifact['id']}",
+        }
+
+        name = f"job-{self._logged_code_artifact['name']}"
+
+        artifact = Artifact(name, JOB_ARTIFACT_TYPE)
+        return artifact, source
+
+    def _build_image_job(
+        self, image_name: str, args: List[str]
+    ) -> Tuple[Artifact, ImageSourceDict]:
+        name = make_artifact_name_safe(f"job-{image_name}")
+        artifact = Artifact(name, JOB_ARTIFACT_TYPE)
+        source: ImageSourceDict = {
+            "image": image_name,
+            "args": args,
+        }
+        return artifact, source
+
+    def build(self) -> Optional[Artifact]:
+        if not os.path.exists(
+            os.path.join(self._settings.files_dir, REQUIREMENTS_FNAME)
+        ):
+            return None
+        metadata = self._handle_metadata_file()
+        if metadata is None:
+            return None
+        git_info: Dict[str, str] = metadata.get("git", {})
+        image_name: Optional[str] = metadata.get("docker", None)
+        program_relpath: Optional[str] = metadata.get("codePath", None)
+        args: List[str] = metadata.get("args", [])
+        runtime: Optional[str] = metadata.get("python", None)
+        # can't build a job without a python version
+        if runtime is None:
+            return None
+        artifact = None
+        source_type = None
+        source: Optional[
+            Union[GitSourceDict, ArtifactSourceDict, ImageSourceDict]
+        ] = None
+        if self._has_git_job_ingredients(git_info, program_relpath):
+            remote = git_info.get("remote")
+            commit = git_info.get("commit")
+            assert isinstance(remote, str)
+            assert isinstance(commit, str)
+            assert program_relpath is not None
+            artifact, source = self._build_repo_job(
+                remote, commit, program_relpath, args
+            )
+            source_type = "repo"
```

Comment:
should all of this be part of the build?

Reply:
Moved into build